untyped-lambda-calculus/article.md

```diff
@@ -0,0 +1,552 @@
+
+We introduce new abstractions using the $\lambda$ symbol.
+An abstraction consists of a _head_ and a _body_, separated by the dot ($.$) symbol.
+The head contains the $\lambda$ symbol and the parameter name.
+The body is an arbitrary expression.
+For example, $\lambda x. x$ is an abstraction.
+
+Lambda calculus derives its name from the $\lambda$ symbol used in this notation.
+The symbol was most likely chosen arbitrarily.
+However, there's an apocryphal story that the notation was derived from $\hat x$ morphing first into $\Lambda x$ and later into $\lambda x$.
+
+Variables in the body of an abstraction that are parameters of said abstraction are called _bound variables_.
+Other variables are called _free variables_.
+
+For example, consider the expression:
+
+$\lambda y. (\lambda z. x\;y\;z)\;z.$
+
+Variables $y,$ $z$ in the body of the inner abstraction $(x\;y\;z)$ are bound in this expression.
+Variable $x$ is free everywhere.
+The last variable $z$ in the outer abstraction $\lambda y. \ldots\;z$ is also free, because no abstraction binds (i.e. introduces) it before it is used.
+All bindings are local.
+
+An expression without free variables is called a _closed term_ or a _combinator_.
+
+For example, the expression $\lambda x.\lambda y.\lambda z. x\;y\;z.$ is a combinator because $x$, $y$, and $z$ are bound in one of the outer (relative to their use) abstractions.
+However, the sub-expression $\lambda y.\lambda z. x\;y\;z$ _by itself_ is not.
+
+The choice of bound variable names is arbitrary.
+Expressions that differ only in the names of bound variables are called _alpha-equivalent_.
+For example, $\lambda x. x$ and $\lambda y. y$ are alpha-equivalent, but $\lambda x. y$ and $\lambda y. z$ are not.
+
+And, as usual in abstract mathematics, we can assign a name to an expression.
+Here we'll be using the equals sign ($=$) for that.
+For example: $id = \lambda x. x.$
+
+### Applications
+
+Application is the operation of substituting a specific value for the parameter in the abstraction.
+It's the only operation defined in the pure untyped lambda calculus.
+
+In the standard notation, application is denoted simply by whitespace.
+For example, $f\; x$ is the application of $f$ to $x$.
+Since all abstractions are univariate by definition, there's no special syntax for multivariate application.
+
+You might notice that using whitespace can lead to ambiguity.
+We'll use parentheses where necessary to avoid it.
+Additionally, we will assume that application is left-associative and binds more tightly than abstractions.
+Hence, $f\;a\;b\;c$ is $((f\;a)\;b)\;c$, i.e. $f$ is applied to $a$, then the result is applied to $b$, then to $c$.
+Since application binds more tightly, we don't need parentheses around the body of an abstraction.
+However, we might need parenthesis around the whole abstraction itself.
+We will also eschew parentheses for nested abstractions, e.g. $\lambda x. \lambda y. x\;y$ is $\lambda x. (\lambda y. (x\;y)).$
```

Comment:
The example with both nested abstractions and applications might be helpful as well.

Reply:
I think that's going to be a bit overkill -- the rules are ultimately pretty simple, and getting too caught up in the examples will break the flow.